File: src/main/java/at/yawk/wm/Scheduler.java
package at.yawk.wm;

import java.util.concurrent.Executor;
import java.util.concurrent.Future;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import javax.annotation.Nonnull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

/**
 * Wrapper for {@link ScheduledExecutorService} that logs task exceptions and never fails.
 *
 * @author yawkat
 */
@Slf4j
@RequiredArgsConstructor
public class Scheduler implements Executor {
    // we use a scheduled service for timing and then delegate execution to a cached thread pool
    private final ScheduledExecutorService scheduledService;
    private final Executor immediateService;

    public Future<?> scheduleAtFixedRate(Runnable task, long initialDelay, long interval, TimeUnit unit) {
        return scheduledService.scheduleAtFixedRate(delegating(locked(task)), initialDelay, interval, unit);
    }

    public Future<?> schedule(Runnable task, long delay, TimeUnit unit) {
        return scheduledService.schedule(delegating(task), delay, unit);
    }

    @Override
    public void execute(@Nonnull Runnable task) {
        delegating(task).run();
    }

    private Runnable delegating(@Nonnull Runnable task) {
        return () -> immediateService.execute(() -> {
            try {
                task.run();
            } catch (Throwable t) {
                log.error("Error in task", t);
            }
        });
    }

    private static Runnable locked(Runnable runnable) {
        Lock lock = new ReentrantLock(true);
        return () -> {
            lock.lock();
            try {
                runnable.run();
            } finally {
                lock.unlock();
            }
        };
    }
}

File: src/main/java/at/yawk/wm/Scheduler.kt
package at.yawk.wm

import java.util.concurrent.Executor
import java.util.concurrent.Future
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock

private fun locked(runnable: Runnable): Runnable {
    val lock = ReentrantLock(true)
    return Runnable {
        lock.lock()
        try {
            runnable.run()
        } finally {
            lock.unlock()
        }
    }
}

/**
 * Wrapper for [ScheduledExecutorService] that logs task exceptions and never fails.

 * @author yawkat
 */
class Scheduler(
        // we use a scheduled service for timing and then delegate execution to a cached thread pool
        private val scheduledService: ScheduledExecutorService,
        private val immediateService: Executor
) : Executor {

    fun scheduleAtFixedRate(task: Runnable, initialDelay: Long, interval: Long, unit: TimeUnit): Future<*> {
        return scheduledService.scheduleAtFixedRate(delegating(locked(task)), initialDelay, interval, unit)
    }

    fun schedule(task: Runnable, delay: Long, unit: TimeUnit): Future<*> {
        return scheduledService.schedule(delegating(task), delay, unit)
    }

    override fun execute(task: Runnable) {
        delegating(task).run()
    }

    private fun delegating(task: Runnable): Runnable {
        return Runnable {
            immediateService.execute {
                try {
                    task.run()
                } catch (t: Throwable) {
                    log.error("Error in task", t)
                }
            }
        }
    }

    companion object {
        private val log = org.slf4j.LoggerFactory.getLogger(Scheduler::class.java)
    }
}
